app/src/main/java/com/fireblocks/sdkdemo/FireblocksManager.kt

```diff
@@ -633,6 +661,62 @@ class FireblocksManager : CoroutineScope {
         }
     }
 
+    fun getOrCreatePassphraseId(context: Context, passphraseLocation: PassphraseLocation, callback: (String?) -> Unit){
+        launch {
+            withContext(Dispatchers.IO) {
+                getPassphraseId(context) { passphraseId ->
+                    if (passphraseId.isNullOrEmpty()) {
+                        val generatedPassphraseId  = Fireblocks.generatePassphraseId()
+                        createPassphraseInfo(context, generatedPassphraseId, passphraseLocation) { success ->
+                            if (success) {
+                                callback(generatedPassphraseId)
+                            } else {
+                                Timber.e("Failed to createPassphraseInfo")
+                                callback(null)
+                            }
+                        }
+                    } else {
+                        callback(passphraseId)
+                    }
+                }
+            }
+        }
+    }
+
+    private fun getPassphraseId(context: Context, callback: (String?) -> Unit) {
+        var passphraseId: String? = null
+        runCatching {
+            val deviceId = getDeviceId()
+            val response = Api.with(StorageManager.get(context, deviceId)).getPassphraseInfos().execute()
+            Timber.d("got response from getPassphraseInfos rest API, code:${response.code()}, isSuccessful:${response.isSuccessful} response.body(): ${response.body()}",
+                response)
+            passphraseId = response.body()?.let { passphraseInfos ->
+                if (!passphraseInfos.passphrases.isNullOrEmpty()) {
+                    passphraseInfos.passphrases.last().passphraseId
+                } else {
+                    null
+                }
+            }
+        }.onFailure {
+            Timber.e(it, "Failed to call getPassphraseInfos API")
+        }
+        callback.invoke(passphraseId)
+    }
+
+    private fun createPassphraseInfo(context: Context, passphraseId: String, passphraseLocation: PassphraseLocation, callback: (success: Boolean) -> Unit) {
+        var success = false
+        runCatching {
+            val deviceId = getDeviceId()
+            val response = Api.with(StorageManager.get(context, deviceId)).createPassphraseInfo(passphraseId, body = PassphraseInfo(location = passphraseLocation)).execute()
+            val body = response.body()
+            success = response.isSuccessful
+            Timber.d("got response from createPassphraseInfo rest API code:${response.code()}, isSuccessful:${response.isSuccessful} response.body(): $body", response)
```

Comment:
you should create a method that does this, you have several of these
something along the lines "logResponse(api:"", response:response)"

Reply:
ok